### PR TITLE
Adding fixes and workarounds to allow compilation with nvcc/msvc (VS2015up3)

### PR DESCRIPTION
**Pull request by hkaiser**
This also fixes some possible problems with base_lco_with_value

### CODE REVIEW

File: hpx/lcos/server/channel.hpp [hunk 1]
```diff
@@ -75,7 +75,11 @@ namespace hpx { namespace lcos { namespace server
...
 
         // Retrieve the next value from the channel
-        result_type get_value(error_code& ec = throws)
+        result_type get_value()
+        {
+            return channel_.get(launch::sync);
+        }
+        result_type get_value(error_code& ec)
         {
             return channel_.get(launch::sync);
```
Comment (sithhell) [inline hpx/lcos/server/channel.hpp]: Did you miss to pass the error code here?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1088,15 +1088,15 @@ if(HPX_WITH_CUDA AND NOT HPX_WITH_CUDA_CLANG)
     hpx_library_dir(${CUDA_TOOLKIT_ROOT_DIR}/lib/x64)
     link_directories(${CUDA_TOOLKIT_ROOT_DIR}/lib/x64)
     set(CUDA_NVCC_FLAGS_DEBUG ${CUDA_NVCC_FLAGS_DEBUG};-D_DEBUG;-O0;-g;-G;-Xcompiler=-MDd;-Xcompiler=-Od;-Xcompiler=-Zi;-Xcompiler=-bigobj)
-    set(CUDA_NVCC_FLAGS_RELWITHDEBINFO ${CUDA_NVCC_FLAGS_RELWITHDEBINFO};-DNDEBUG;-O2;-g;-Xcompiler=-MD,-O2,-Zi;-Xcompiler=-bigobj)
+    set(CUDA_NVCC_FLAGS_RELWITHDEBINFO ${CUDA_NVCC_FLAGS_RELWITHDEBINFO};-DNDEBUG;-O2;-g;--verbose;-Xcompiler=-MD,-O2,-Zi;-Xcompiler=-bigobj;--keep)
```
Comment (hkaiser) [inline CMakeLists.txt]: I don't see any other way as nvcc is not picking up the flags set through cmake generator settings. I had to comment out this for instance: https://github.com/STEllAR-GROUP/hpx/blob/3f87b2da11b02ff327d234adf5634d730a695bda/cmake/HPX_AddCompileFlag.cmake#L33.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1088,15 +1088,15 @@ if(HPX_WITH_CUDA AND NOT HPX_WITH_CUDA_CLANG)
...
     set(CUDA_NVCC_FLAGS_MINSIZEREL ${CUDA_NVCC_FLAGS_MINSIZEREL};-DNDEBUG;-O1;-Xcompiler=-MD,-O1;-Xcompiler=-bigobj)
     set(CUDA_NVCC_FLAGS_RELEASE ${CUDA_NVCC_FLAGS_RELEASE};-DNDEBUG;-O2;-Xcompiler=-MD,-Ox;-Xcompiler=-bigobj)
   endif()
   set(CUDA_SEPARABLE_COMPILATION ON)
   set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-gencode=arch=compute_30,code=sm_30)
-  set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-gencode=arch=compute_35,code=sm_35)
+#  set(CUDA_NVCC_FLAGS ${CUDA_NVCC_FLAGS};-gencode=arch=compute_35,code=sm_35)
```
Comment (hkaiser) [inline CMakeLists.txt]: Sorry, that's a left-over from experimentation

File: CMakeLists.txt (hunk 2 above)
Comment (sithhell) [inline CMakeLists.txt]: Hmm, I don't like having that as a default. Sounds like there's a bigger underlying issue here. Could we open a ticket and fix it properly instead?

Comment (hkaiser) [inline CMakeLists.txt]: Sure. You have 'solved' the very same issue by unconditionally setting the compilation flags for all build configurations (here: https://github.com/STEllAR-GROUP/hpx/blob/3f87b2da11b02ff327d234adf5634d730a695bda/cmake/HPX_AddCompileFlag.cmake#L34). This however does not work for multi-target build systems like VS. So while I would like to solve it, I have no idea how to approach this issue as it seems to be a problem in FindCUDA.cmake not picking up settings from cmake generator constructs.

**Comment by sithhell** [comment]
We should have really reverted the nvcc flags and look for a proper fix.
Not everyone wants to keep the intermediates and have verbose output.

Am 07.12.2016 11:17 nachm. schrieb "Hartmut Kaiser" <
notifications@github.com>:

**Comment by hkaiser** [comment]
I will remove `--verbose` and `--keep`

**Pull request merged**